src/GafferSceneUI/LightPositionTool.cpp

```diff
@@ -0,0 +1,347 @@
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferSceneUI/LightPositionTool.h"
+#include "GafferSceneUI/SceneView.h"
+
+#include "GafferSceneUI/ContextAlgo.h"
+
+#include "Gaffer/MetadataAlgo.h"
+
+#include "IECore/AngleConversion.h"
+
+IECORE_PUSH_DEFAULT_VISIBILITY
+#include "OpenEXR/OpenEXRConfig.h"
+#if OPENEXR_VERSION_MAJOR < 3
+#include "OpenEXR/ImathEuler.h"
+#include "OpenEXR/ImathMatrixAlgo.h"
+#else
+#include "Imath/ImathEuler.h"
+#include "Imath/ImathMatrixAlgo.h"
+#endif
+IECORE_POP_DEFAULT_VISIBILITY
+
+#include "boost/algorithm/string/predicate.hpp"
+#include "boost/bind/bind.hpp"
+
+#include "fmt/format.h"
+
+using namespace boost::placeholders;
+using namespace Imath;
+using namespace Gaffer;
+using namespace GafferUI;
+using namespace GafferScene;
+using namespace GafferSceneUI;
+
+GAFFER_NODE_DEFINE_TYPE( LightPositionTool );
+
+LightPositionTool::ToolDescription<LightPositionTool, SceneView> LightPositionTool::g_toolDescription;
+size_t LightPositionTool::g_firstPlugIndex = 0;
+
+LightPositionTool::LightPositionTool( SceneView *view, const std::string &name ) :
+	TransformTool( view, name ),
+	m_shadowPivot( std::nullopt ),
+	m_shadowPoint( std::nullopt )
+{
+	SceneGadget *sg = runTimeCast<SceneGadget>( this->view()->viewportGadget()->getPrimaryChild() );
+	// We have to insert this before the underlying SelectionTool connections or it starts an object drag.
+	sg->buttonPressSignal().connectFront( boost::bind( &LightPositionTool::buttonPress, this, ::_2 ) );
+
+	plugSetSignal().connect( boost::bind( &LightPositionTool::plugSet, this, ::_1 ) );
+
+	connectToViewContext();
+	view->contextChangedSignal().connect( boost::bind( &LightPositionTool::connectToViewContext, this  ) );
+
+	view->plugInputChangedSignal().connect( boost::bind( &LightPositionTool::plugInputChanged, this, ::_1 ) );
+
+	selectionChangedSignal().connect( boost::bind( &LightPositionTool::selectionChanged, this, ::_1 ) );
+
+	storeIndexOfNextChild( g_firstPlugIndex );
+}
+
+LightPositionTool::~LightPositionTool()
+{
+}
+
+void LightPositionTool::position( const V3f &shadowPivot, const V3f &shadowPoint )
+{
+	const V3f newCentroid =
+		( shadowPivot - shadowPoint ).normalized() *
+		( m_worldCentroid - shadowPivot ).length() +
+		shadowPivot
+	;
+
+	const M44f centroidTransform = rotationMatrix( m_worldCentroid - shadowPivot, newCentroid - shadowPivot );
+
+	UndoScope undoScope( selection().back().editTarget()->ancestor<ScriptNode>() );
+
+	for( const auto &s : selection() )
+	{
+		// See `RotateTool::buttonPress()` for a description of why we use this relatively
+		// elaborate orientation calculation.
```

Comment:
I'm not sure the orientation stuff is working great in this context. I tested with a spotlight with a gobo, and although the initial click keeps the gobo upright, after a while it has drifted off so it's rotated on its side. You can see the same thing with a quadlight - use a second viewer to look through the quadlight while positioning it in the other viewer, and you'll see the look-through get a twist on it.

I'm not sure what the behaviour should be if the light isn't upright (local Y aligned-ish with world Y) in the first place, but if it _is_ upright it feels to me that it should remain that way (i.e. local Y is as aligned to world Y as it can be while still achieving the required aim).

Reply:
I'm still finding this to be the case in the latest version.